functional/otlp/func_grpc_main.cc

```diff
@@ -0,0 +1,821 @@
+  std::string m_name;
+  test_func_t m_func;
+};
+
+int test_basic();
+
+int test_cert_not_found();
+int test_cert_invalid();
+int test_cert_unreadable();
+int test_client_cert_not_found();
+int test_client_cert_invalid();
+int test_client_cert_unreadable();
+int test_client_cert_no_key();
+int test_client_key_not_found();
+int test_client_key_invalid();
+int test_client_key_unreadable();
+
+int test_mtls_ok();
+
+static const test_case all_tests[] = {{"basic", test_basic},
+                                      {"cert-not-found", test_cert_not_found},
+                                      {"cert-invalid", test_cert_invalid},
+                                      {"cert-unreadable", test_cert_unreadable},
+#ifdef ENABLE_OTLP_GRPC_SSL_MTLS_PREVIEW
+                                      {"client-cert-not-found", test_client_cert_not_found},
+                                      {"client-cert-invalid", test_client_cert_invalid},
+                                      {"client-cert-unreadable", test_client_cert_unreadable},
+                                      {"client-cert-no-key", test_client_cert_no_key},
+                                      {"client-key-not-found", test_client_key_not_found},
+                                      {"client-key-invalid", test_client_key_invalid},
+                                      {"client-key-unreadable", test_client_key_unreadable},
+                                      {"mtls-ok", test_mtls_ok},
+#endif  // ENABLE_OTLP_GRPC_SSL_MTLS_PREVIEW
+                                      {"", nullptr}};
+
+void list_test_cases()
+{
+  const test_case *current = all_tests;
+
+  while (current->m_func != nullptr)
+  {
+    std::cout << current->m_name << '\n';
+    current++;
+  }
+}
+
+int run_test_case(const std::string &name)
+{
+  const test_case *current = all_tests;
+
+  while (current->m_func != nullptr)
+  {
+    if (current->m_name == name)
+    {
+      int rc = current->m_func();
+      return rc;
+    }
+    current++;
+  }
+
+  std::cerr << "Unknown test <" << name << ">" << '\n';
+  return 1;
+}
+
+int main(int argc, char *argv[])
+{
+  // Program name
+  argc--;
+  argv++;
+
+  int rc = parse_args(argc, argv);
+
+  if (rc != 0)
+  {
+    usage(stderr);
+    return 1;
+  }
+
+  if (opt_help)
+  {
+    usage(stdout);
+    return 0;
+  }
+
+  if (opt_list)
+  {
+    list_test_cases();
+    return 0;
+  }
+
+  if (opt_endpoint.find("https:") != std::string::npos)
+  {
+    opt_secure = true;
+  }
+  else
+  {
+    opt_secure = false;
+  }
+
+  opentelemetry::nostd::shared_ptr<internal_log::LogHandler> log_handler(new TestLogHandler);
+
+  internal_log::GlobalLogHandler::SetLogHandler(log_handler);
+  internal_log::GlobalLogHandler::SetLogLevel(internal_log::LogLevel::Debug);
+
+  rc = run_test_case(opt_test_name);
+  return rc;
+}
+
+void set_common_opts(otlp::OtlpGrpcExporterOptions &opts)
+{
+  opts.endpoint            = opt_endpoint;
+  opts.timeout             = std::chrono::milliseconds{100};
```

Comment:
Setting a timeout to speed things up a bit, since some tests might go on for 5+ seconds before erroring out.
The observable behavior is that a "deadline exceeded" will be output as a result instead of the actual reason. This is accounted for in `expect_connection_failed()`.

If deemed more appropriate, it could be reverted so as to let the tests fail normally.